api_tests/nodes/views/test_node_contributors_list.py

```diff
@@ -811,15 +804,263 @@ def test_adds_contributor_private_project_not_logged_in(self):
+
+    def test_add_contributor_validation_user_id(self):
+        self.validate_data(NodeContributorsCreateSerializer(), user_id='abcde')
+
+    def test_add_contributor_validation_user_id_fullname(self):
+        with assert_raises(Conflict):
+            self.validate_data(NodeContributorsCreateSerializer(), user_id='abcde', full_name='Kanye')
+
+    def test_add_contributor_validation_user_id_email(self):
+        with assert_raises(Conflict):
+            self.validate_data(NodeContributorsCreateSerializer(), user_id='abcde', email='kanye@west.com')
+
+    def test_add_contributor_validation_user_id_fullname_email(self):
+        with assert_raises(Conflict):
+            self.validate_data(NodeContributorsCreateSerializer(), user_id='abcde', full_name='Kanye', email='kanye@west.com')
+
+    def test_add_contributor_validation_fullname(self):
+        self.validate_data(NodeContributorsCreateSerializer(), full_name='Kanye')
+
+    def test_add_contributor_validation_email(self):
+        with assert_raises(exceptions.ValidationError):
+            self.validate_data(NodeContributorsCreateSerializer(), email='kanye@west.com')
+
+    def test_add_contributor_validation_fullname_email(self):
+        self.validate_data(NodeContributorsCreateSerializer(), full_name='Kanye', email='kanye@west.com')
+
+
+class TestNodeContributorCreateEmail(NodeCRUDTestCase):
+
+    def setUp(self):
+        super(TestNodeContributorCreateEmail, self).setUp()
+        self.url = '/{}nodes/{}/contributors/'.format(API_BASE, self.public_project._id)
+
+    @mock.patch('framework.auth.views.mails.send_mail')
+    def test_add_contributor_no_email_if_false(self, mock_mail):
+        url = '{}?send_email=false'.format(self.url)
+        payload = {
+            'data': {
+                'type': 'contributors',
+                'attributes': {
+                    'full_name': 'Kanye West',
+                    'email': 'kanye@west.com'
+                }
+            }
+        }
+        res = self.app.post_json_api(url, payload, auth=self.user.auth)
+        assert_equal(res.status_code, 201)
+        assert_equal(mock_mail.call_count, 0)
+
+    @mock.patch('framework.auth.views.mails.send_mail')
+    def test_add_contributor_email_if_default(self, mock_mail):
+        url = '{}?send_email=default'.format(self.url)
+        payload = {
+            'data': {
+                'type': 'contributors',
+                'attributes': {
+                },
+                'relationships': {
+                    'users': {
+                        'data': {
+                            'type': 'users',
+                            'id': self.user_two._id
+                        }
+                    }
+                }
+            }
+        }
+
+        with capture_signals() as mock_signal:
+            res = self.app.post_json_api(url, payload, auth=self.user.auth)
+        assert_in(contributor_added, mock_signal.signals_sent())
```

Comment:
You miiiight be able to use assert_called_with to ensure that the signal was sent with the right info. https://docs.python.org/3/library/unittest.mock.html#quick-guide  Worth a quick check.